### PR TITLE
Add Queue Retry Command to Appwrite

### PR DESCRIPTION
**Pull request by PineappleIOnic**
## What does this PR do?

This PR adds a command that will retry queues in Appwrite

## Test Plan

Create a couple failure jobs, run the command and see if the worker attempts them again

## Checklist

- [X] Have you read the [Contributing Guidelines on issues](https://github.com/appwrite/appwrite/blob/master/CONTRIBUTING.md)?
- [ ] If the PR includes a change to an API's metadata (desc, label, params, etc.), does it also include updated API specs and example docs?

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -85,6 +85,7 @@ RUN chmod +x /usr/local/bin/doctor && \
     chmod +x /usr/local/bin/ssl && \
     chmod +x /usr/local/bin/test && \
     chmod +x /usr/local/bin/vars && \
+    chmod +x /usr/local/bin/retry && \
```
Comment (eldadfux) [inline Dockerfile:88]: Maybe we can have a prefix for queue related tasks. Not sure how complex `queue retry` will be, but we can also consider `queue-retry`, I would also consider a few command to give the size of the queue, and size of failed jobs.

Maybe

`queue-count-jobs`
`queue-count-failures`

WDYT?

File: composer.json [hunk 2]
```diff
@@ -62,7 +62,7 @@
         "utopia-php/platform": "0.5.*",
         "utopia-php/pools": "0.4.*",
         "utopia-php/preloader": "0.2.*",
-        "utopia-php/queue": "0.5.*",
+        "utopia-php/queue": "dev-feat-retry",
```
Comment (eldadfux) [inline composer.json:65]: Reminder to update the tag.

File: src/Appwrite/Platform/Tasks/Retry.php [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+            ->inject('queue')
+            ->callback(fn ($queueName, $queue) => $this->action($queueName, $queue));
+    }
+
+    public function action(string $queueName, Connection $queue): void
+    {
+        $queueClient = new Client($queueName, $queue);
+
+        if ($queueClient->sumFailedJobs() === 0) {
+            Console::error('Found no jobs to retry, are you sure you have the right queue name?');
```
Comment (eldadfux) [inline src/Appwrite/Platform/Tasks/Retry.php:33]:
```suggestion
            Console::error('No failed jobs found.');
```

File: src/Appwrite/Platform/Tasks/Retry.php [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+            ->param('name', '', new Text(128), 'Queue name')
+            ->inject('queue')
+            ->callback(fn ($queueName, $queue) => $this->action($queueName, $queue));
+    }
+
+    public function action(string $queueName, Connection $queue): void
+    {
+        $queueClient = new Client($queueName, $queue);
+
+        if ($queueClient->sumFailedJobs() === 0) {
```
Comment (eldadfux) [inline src/Appwrite/Platform/Tasks/Retry.php:32]: I might call this method `countFailedJobs `

**Review by eldadfux — CHANGES_REQUESTED** [review]

**Comment by eldadfux** [comment]
@PineappleIOnic do we also have a good place where we can start documenting these CLI commands better?

File: src/Appwrite/Platform/Tasks/Retry.php (hunk 4 above)
Comment (PineappleIOnic) [reply src/Appwrite/Platform/Tasks/Retry.php:32]: This would need to be updated in the Queue lib

Comment (PineappleIOnic) [reply src/Appwrite/Platform/Tasks/Retry.php:32]: I've updated all sumX methods to countX methods

File: src/Appwrite/Platform/Tasks/Retry.php [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+    public static function getName(): string
+    {
+        return 'retry';
+    }
+
+
+    public function __construct()
+    {
+        $this
+            ->desc('Retry Queue')
```
Comment (christyjacob4) [inline src/Appwrite/Platform/Tasks/Retry.php:22]: lets have a better description "Retry failed jobs from a specific queue identified by the name parameter"

File: src/Appwrite/Platform/Tasks/Retry.php [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+            ->param('name', '', new Text(128), 'Queue name')
```
Comment (christyjacob4) [inline src/Appwrite/Platform/Tasks/Retry.php:23]: name here should use the whitelist validator

File: Dockerfile [hunk 7]
```diff
@@ -85,6 +85,7 @@ RUN chmod +x /usr/local/bin/doctor && \
     chmod +x /usr/local/bin/ssl && \
     chmod +x /usr/local/bin/test && \
     chmod +x /usr/local/bin/vars && \
+    chmod +x /usr/local/bin/queue-retry && \
```
Comment (christyjacob4) [inline Dockerfile:88]:
```suggestion
    chmod +x /usr/local/bin/retry-jobs && \
```

File: bin/queue-retry [hunk 8]
```diff
@@ -0,0 +1,3 @@
+#!/bin/sh
+
+php /usr/src/code/app/cli.php retry $@
```
Comment (christyjacob4) [inline bin/queue-retry:3]:
```suggestion
php /usr/src/code/app/cli.php retry-jobs $@
```

File: src/Appwrite/Platform/Services/Tasks.php [hunk 9]
```diff
@@ -46,6 +47,7 @@ public function __construct()
             ->addAction(DeleteOrphanedProjects::getName(), new DeleteOrphanedProjects())
             ->addAction(PatchRecreateRepositoriesDocuments::getName(), new PatchRecreateRepositoriesDocuments())
             ->addAction(GetMigrationStats::getName(), new GetMigrationStats())
+            ->addAction(Retry::getName(), new Retry())
```
Comment (christyjacob4) [inline src/Appwrite/Platform/Services/Tasks.php:50]:
```suggestion
            ->addAction(Retry::getName(), new RetryJobs())
```

File: src/Appwrite/Platform/Tasks/Retry.php [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+            ->inject('queue')
+            ->callback(fn ($queueName, $queue) => $this->action($queueName, $queue));
```
Comment (christyjacob4) [inline src/Appwrite/Platform/Tasks/Retry.php:25]:
```suggestion
            ->callback(fn ($name, $queue) => $this->action($name, $queue));
```

File: src/Appwrite/Platform/Tasks/Retry.php [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+    }
+
+    public function action(string $queueName, Connection $queue): void
```
Comment (christyjacob4) [inline src/Appwrite/Platform/Tasks/Retry.php:28]:
```suggestion
    public function action(string $name, Connection $queue): void
```
Lets also add a doc block explaining the function and the parameters

File: src/Appwrite/Platform/Tasks/Retry.php [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+    {
+        $queueClient = new Client($queueName, $queue);
```
Comment (christyjacob4) [inline src/Appwrite/Platform/Tasks/Retry.php:30]: If the `name` is empty, we must log an error and return..

File: src/Appwrite/Platform/Services/Tasks.php [hunk 13]
```diff
@@ -21,6 +21,7 @@
 use Appwrite\Platform\Tasks\DeleteOrphanedProjects;
 use Appwrite\Platform\Tasks\GetMigrationStats;
 use Appwrite\Platform\Tasks\PatchRecreateRepositoriesDocuments;
+use Appwrite\Platform\Tasks\RetryJobs;
```
Comment (christyjacob4) [inline src/Appwrite/Platform/Services/Tasks.php:24]:
```suggestion
use Appwrite\Platform\Tasks\QueueRetry;
```

File: src/Appwrite/Platform/Tasks/RetryJobs.php [hunk 14]
```diff
@@ -0,0 +1,48 @@
...
+use Utopia\Platform\Action;
+use Utopia\Queue\Client;
+use Utopia\Queue\Connection;
+use Utopia\Validator\Text;
+
+class RetryJobs extends Action
+{
+    public static function getName(): string
+    {
+        return 'retry-jobs';
```
Comment (christyjacob4) [inline src/Appwrite/Platform/Tasks/RetryJobs.php:15]:
```suggestion
        return 'queue-retry';
```

File: src/Appwrite/Platform/Tasks/RetryJobs.php [hunk 15]
```diff
@@ -0,0 +1,48 @@
...
+            return;
+        }
+
+        $queueClient = new Client($name, $queue);
+
+        if ($queueClient->countFailedJobs() === 0) {
+            Console::error('No failed jobs found.');
+            return;
+        }
+
```
Comment (christyjacob4) [inline src/Appwrite/Platform/Tasks/RetryJobs.php:45]: Add a log here to indicate that jobs are being retried

File: src/Appwrite/Platform/Tasks/RetryJobs.php [hunk 16]
```diff
@@ -0,0 +1,48 @@
...
+     * @param Connection $queue
+     */
+    public function action(string $name, Connection $queue): void
+    {
+        if (!$name) {
+            Console::error('Missing required parameter $name');
+            return;
+        }
+
+        $queueClient = new Client($name, $queue);
```
Comment (christyjacob4) [inline src/Appwrite/Platform/Tasks/RetryJobs.php:39]: We have to validate if name is a legitimate queue name... 

Use the WhiteList validator for the `name` param above and pass Whitelist([Event::FUNCTIONS_QUEUE_NAME, etc.])

**Review by christyjacob4 — APPROVED** [review]

**Comment by christyjacob4** [comment]
@PineappleIOnic the database tests are failing.. Please check whats going on and also change the base branch to `refactor-usage-sn`

**Comment by PineappleIOnic** [comment]
It looks like the database error messages were changed without updating the tests, I've gone ahead and updated those tests with the new error messages

**Comment by PineappleIOnic** [comment]
They all use `getQueueSize` not any count or sum methods, so they do not need to be updated

**Review by christyjacob4 — APPROVED** [review]

**Pull request merged**